nixos/modules/services/networking/netbird.nix

```diff
@@ -1,171 +1,496 @@
+      '';
+      example = lib.literalExpression ''
+        {
+          services.netbird.clients.wt0.port = 51820;
+          services.netbird.clients.personal.port = 51821;
+          services.netbird.clients.work1.port = 51822;
+        }
       '';
     };
   };
 
   config = mkMerge [
     (mkIf cfg.enable {
-      # For backwards compatibility
-      services.netbird.tunnels.wt0.stateDir = "netbird";
+      services.netbird.clients.wt0 = {
+        port = mkDefault 51820;
+        name = mkDefault "netbird";
+        interface = mkDefault "wt0";
+        hardened = mkDefault false;
+      };
     })
+    {
+      boot.extraModulePackages = optional
+        (cfg.clients != { } && (versionOlder kernel.version "5.6"))
+        kernelPackages.wireguard;
 
-    (mkIf (cfg.tunnels != { }) {
-      boot.extraModulePackages = optional (versionOlder kernel.kernel.version "5.6") kernel.wireguard;
+      environment.systemPackages =
+        toClientList (client: client.wrapper)
+        # omitted due to https://github.com/netbirdio/netbird/issues/1562
+        #++ optional (cfg.clients != { }) cfg.package
+        # omitted due to https://github.com/netbirdio/netbird/issues/1581
+        #++ optional (cfg.clients != { } && cfg.ui.enable) cfg.ui.package
+      ;
 
-      environment.systemPackages = [ cfg.package ];
+      networking.dhcpcd.denyInterfaces = toClientList (client: client.interface);
+      networking.networkmanager.unmanaged = toClientList (client: "interface-name:${client.interface}");
 
-      networking.dhcpcd.denyInterfaces = attrNames cfg.tunnels;
+      networking.firewall.allowedUDPPorts = concatLists (toClientList (client: optional client.openFirewall client.port));
 
-      systemd.network.networks = mkIf config.networking.useNetworkd (
-        mapAttrs'
-          (
-            name: _:
-            nameValuePair "50-netbird-${name}" {
-              matchConfig = {
-                Name = name;
-              };
-              linkConfig = {
-                Unmanaged = true;
-                ActivationPolicy = "manual";
-              };
-            }
-          )
-          cfg.tunnels
-      );
+      systemd.network.networks = mkIf config.networking.useNetworkd (toClientAttrs (client:
+        nameValuePair "50-netbird-${client.interface}" {
+          matchConfig = {
+            Name = client.interface;
+          };
+          linkConfig = {
+            Unmanaged = true;
+            ActivationPolicy = "manual";
+          };
+        }
+      ));
 
-      systemd.services =
-        mapAttrs'
-          (
-            name:
-            { environment, stateDir, ... }:
-            nameValuePair "netbird-${name}" {
-              description = "A WireGuard-based mesh network that connects your devices into a single private network";
+      environment.etc = toClientAttrs (client: nameValuePair "${client.name}/config.d/50-nixos.json" {
```

Comment:
Why not use `builtins.toJSON client.config` expression directly in the `preStart` script?

Reply:
I want to be able to add additional configs (eg credentials) outside the NixOS configuration

Reply:
It's fine to add a workaround until the issue is fixed but I don't think it's a good idea to advertise ways of configuration which are not supported by upstream

Reply:
good idea, made it explicit in `config` option description.

Reply:
As the configuration is a json file, we can use https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/utils.nix#L160 to manage secrets in it without putting them in the nix store